### DISCUSSION TITLE
org.jgroups.tests.RouterStubGet: Exception in thread "NioClient.Acceptor [srv=a.b.c.d:p]" java.util.ConcurrentModificationException

### DISCUSSION THREAD

**Discussion by ah-ping-luk** [belaban/JGroups#823] [General]
I try to run the helper utility for checking healthiness of the clusters (via TUNNEL / gossip router)

java org.jgroups.tests.RouterStubGet -nio true -host $GOSSIP_HOST -port $GOSSIP_PORT -cluster ISPN

for a cluster with 2 or more member, there will be java.util.ConcurrentModificationException.

For example jgroups version 5.2.23:

```
Exception in thread "NioClient.Acceptor [srv=160.200.147.89:30181]" java.util.ConcurrentModificationException
        at java.base/java.util.HashMap$HashIterator.remove(HashMap.java:1611)
        at org.jgroups.blocks.cs.NioBaseServer$Acceptor.run(NioBaseServer.java:154)
        at java.base/java.lang.Thread.run(Thread.java:840)
```

After a look, there is a bug in the concurrent modification of the Iterator<SelectionKey> 
```

Iterator<SelectionKey> it=null;
...
it=selector.selectedKeys().iterator();
...
**it.remove();**  //line that cause issue
```

javadocs state the selectedKeys is not threadsafe: https://docs.oracle.com/javase/8/docs/api/java/nio/channels/Selector.html#selectedKeys--

```
Keys may be removed from, but not directly added to, the selected-key set. Any attempt to add an object to the key set will cause an [UnsupportedOperationException](https://docs.oracle.com/javase/8/docs/api/java/lang/UnsupportedOperationException.html) to be thrown.

The selected-key set is not thread-safe
```

I have a look of the jgroups source code, this issue seems exists in latest jgroups version 5.3.10

Thank you.

**Comment by belaban** [comment]
Thanks, I created https://issues.redhat.com/browse/JGRP-2820

**Comment by belaban** [comment]
Fixed in master (5.4) and backport to 5.3.11
